framework/validators/UniqueValidator.php

```diff
@@ -122,9 +122,15 @@ public function init()
      */
     public function validateAttribute($model, $attribute)
     {
-        /* @var $targetClass ActiveRecordInterface */
-        $targetClass = $this->getTargetClass($model);
         $targetAttribute = $this->targetAttribute === null ? $attribute : $this->targetAttribute;
+        if ($this->skipOnError) {
+            foreach ((array)$targetAttribute as $k => $v) {
+                if ($model->hasErrors(is_int($k) ? $v : $k)) {
```

Comment:
you miss `$model->validate()` x2

Reply:
No, I don't. Rules are fired one after another so it assumes previous rules are applied already. If developer placed that rule before others - it's his problem.

Reply:
@bizley 1. `$skipOnError`
> whether this validation rule should be skipped if the attribute being validated already has some validation error according to some previous rules.

i.e. this property using only for target attribute, but problem in related attributes
2. it's can be other model

Reply:
We are using current model's attribute value in the query.

Reply:
@bizley nope, `$targetAttribute` in model `$targetClass`:
```php
class Player extends ActiveRecord
{
    public function rules()
    {
        return [
           ['color', 'unique', 'targetClass' => PlayerColor::class, 'targetAttribute' => 'name'],
        ];
    }
}
```

Reply:
Hmm, indeed this requires additional check if `targetClass` is provided.

Reply:
@bizley also, you incorrect use `$skipOnError`: validation skipped if `$attribute` already fail previous validation(s), but you check `$targetAttribute`

Reply:
It looks like you have not read the issue this should fix.

Reply:
@bizley you again miss `$targetAttribute = $this->targetAttribute === null ? $attribute : $this->targetAttribute;` i.e. `$targetAttribute` not always same to `$attribute`: `if ($skipOnError && $model->hasErrors($attribute)) {return;}`

Reply:
Yes, I know that it's not always the same.

Reply:
@bizley you right

Reply:
I think the proposed fix is good. We are using the currently validated model to fetch the values [here](https://github.com/yiisoft/yii2/blob/master/framework/validators/UniqueValidator.php#L264) so there is no need to check what is going on in the target class. Since we should not use the current model's attribute's value if it's not valid (like in the related issue) we should check first whether the attribute has errors or not. Of course if the developer sets `skipOnError` to `false` here or there are no prior validation rules for the used target attribute (developer sets rules in the wrong order or doesn't set rules for the target attribute) the check won't work ending possibly in DB error but this is on his own wish.

Reply:
@bizley nope, you still check `$targetAttribute` instead `$attributes` - need override `UniqueValidator::validateAttributes()`.

Reply:
we must check errors in all passed attributes before validate

Reply:
Why? I cannot find any reason for your statement. [validateAttributes](https://github.com/yiisoft/yii2/blob/master/framework/validators/Validator.php#L251) does not change the model and is not overriden by UniqueValidator and UniqueValidator's [validateAttribute](https://github.com/yiisoft/yii2/blob/master/framework/validators/UniqueValidator.php#L128) is not changing the model as well. Please explain.

Reply:
1.`[['claim_id', 'medication_id'], 'unique', 'targetAttribute' => ['claim_id', 'medication_id']],` - first array is `$attributes`, but you check attributes from `$this->targetAttribute` i.e. `'targetAttribute' => ['claim_id', 'medication_id']`


2. `$skipOnError`
> whether this validation rule should be skipped if the attribute being validated already has some validation error according to some previous rules

We need check `$attributes` passing into `UniqueValidator::validateAttributes()`(look to `Validator::validateAttributes()`). If `$attributes` is string or array with one element, then additional checking must be skipped.

Reply:
3. need check "is attribute validation failed" for every input attribute BEFORE validates attributes as "unique"

Reply:
[framework/validators/UniqueValidator.php](https://github.com/yiisoft/yii2/pull/19417/files)

Reply:
As we already agreed all your points are already addressed by validateAttributes of base Validator class, I'm not sure why you mention this again, maybe you forgot that you agreed. If any of the first attributes has errors and there is skipOnError set we skip the rule, all is fine. The problem from the issue is that there might be target attributes that are not present in the first (main) attributes array that are used for the check later on and are not valid which (MENTIONING AGAIN IN THIS ISSUE) is the problem this PR tries to fix.

Anyway, I'm done discussing it, I'm waiting for a review now from the team members.

Reply:
@bizley i'm agree with "no reason check it again"

Reply:
@bizley your PR break BC:
~~~php
class Item extends ActiveRecord
{
    public function attributes()
    {
        return ['id', 'name'];
    }
    public function rules()
    {
        return [
            ['name', 'required'],
            ['name', 'string', 'min' => 2, 'max' => 200],
        ];
    }
}

class Form extends Model
{
    public $itemName;
    public function rules()
    {
        return [
            ['itemName', 'required'],
            ['itemName', 'min' => 2],
            ['itemName', 'string', 'unique', 'targetClass' => Item::className(), 'targetAttribute' => ['itemName' => 'name']],
        ];
    }
}
~~~